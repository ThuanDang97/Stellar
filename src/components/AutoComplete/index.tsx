import React, { memo, useState, useCallback, useRef, useEffect } from 'react'
import { useTheme } from 'styled-components'

// Components
import Image from 'next/image'
import {
  BorderWrapperStyled,
  TextStyled,
  ViewStyled,
} from '@components/styled-components'
import {
  ListItemStyled,
  ListStyled,
} from '@components/styled-components/ListStyled'

// Utils
import { filterDataContainValue, findItemByValue } from '@utils/index'

// Constants
import { LIST_BILLER } from '@constants/index'

// Props type
import {
  AutoCompleteProps,
  EventProps,
} from '@self-types/components/AutoComplete.props'
import Input from '@components/Input'

const AutoComplete = ({ billerItem, setBillerItem }: AutoCompleteProps) => {
  const theme = useTheme()
  const [placeholder, setPlaceholder] = useState('Select Biller')

  const wrapperRef = useRef<HTMLInputElement>(null)

  const [isToggleDropdown, setToggleDropdown] = useState(false)

  /**
   * Handle toggle dropdown event
   */
  const handleToggleDropdown = () => {
    setToggleDropdown(true)
    setPlaceholder('Search By Name')
  }

  useEffect(() => {
    /**
     * Alert if clicked on outside of element
     */
    const handleClickOutside = (event: MouseEvent) => {
      if (
        wrapperRef &&
        wrapperRef.current &&
        !wrapperRef.current.contains(event.target as Node)
      ) {
        setToggleDropdown(false)
        setPlaceholder('Select Biller')
      }
    }
    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [wrapperRef])

  /**
   * Handle change value
   */
  const handleOnChange = (event: EventProps) => {
    const value = event && event.target && event.target.value
    setBillerItem(value as string)
  }

  /**
   * Handle select biller
   */
  const handleSelectItem = useCallback(
    (id: string) => {
      const selectOption = findItemByValue({
        data: LIST_BILLER,
        key: 'id',
        value: id,
      })

      setBillerItem(selectOption?.title as string)
      setToggleDropdown(false)
    },
    [setBillerItem],
  )

  const renderListBiller = () => {
    const result = filterDataContainValue({
      data: LIST_BILLER,
      value: billerItem,
      key: 'title',
    })

    return (
      <ListStyled
        pTop={theme.metrics.dimensions.xs}
        width={theme.metrics.width.full}
        data-testid="select-menu"
        position="absolute"
        bgColor={theme.colors.white}
      >
        {result.map((item) => {
          const onSelectItem = () => {
            handleSelectItem(item.id)
          }

          return (
            <ListItemStyled
              key={`autocomplete-item-${item.id}`}
              width={theme.metrics.width.full}
              onClick={onSelectItem}
            >
              <TextStyled
                display="flex"
                alignItems="center"
                pLeft={theme.metrics.dimensions.sm}
              >
                <Image
                  src={item.src}
                  width={35}
                  height={35}
                  layout="intrinsic"
                />
                <TextStyled
                  fontSize={theme.typography.fontSize.base}
                  mLeft={theme.metrics.dimensions.base}
                >
                  {item.title}
                </TextStyled>
              </TextStyled>
            </ListItemStyled>
          )
        })}
      </ListStyled>
    )
  }

  return (
    <ViewStyled
      position="relative"
      width={theme.metrics.width.full}
      ref={wrapperRef}
    >
      <BorderWrapperStyled
        data-testid="select"
        width={theme.metrics.width.full}
        onClick={handleToggleDropdown}
      >
        <Input
          type="text"
          iconUrl="/icons/search"
          placeholder={placeholder}
          onChange={handleOnChange}
          value={billerItem}
        />
      </BorderWrapperStyled>
      {isToggleDropdown && renderListBiller()}
    </ViewStyled>
  )
}

export default memo(AutoComplete)
