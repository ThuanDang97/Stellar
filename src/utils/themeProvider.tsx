import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
