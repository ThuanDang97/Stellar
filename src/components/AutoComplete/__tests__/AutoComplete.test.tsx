import '@testing-library/jest-dom'
import { fireEvent, screen } from '@testing-library/react'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

// Component
import AutoComplete from '@components/AutoComplete'

describe('AutoComplete Component', () => {
  const component = <AutoComplete billerItem="" setBillerItem={jest.fn()} />
  it('renders AutoComplete component', () => {
    const { container } = renderWithTheme(component)
    expect(container).toMatchSnapshot()
  })

  test('Render list biller after click on dropdown', () => {
    renderWithTheme(component)

    fireEvent.click(screen.getByTestId('select'))
    expect(screen.getByTestId('select-menu')).toBeInTheDocument()
  })
})
