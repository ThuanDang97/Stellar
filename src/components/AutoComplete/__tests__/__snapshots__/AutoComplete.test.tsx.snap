// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete Component renders AutoComplete component 1`] = `
<div>
  <div
    class="ViewStyled-sc-rqe5sa-0 frRnNW"
    width="100%"
  >
    <div
      class="ViewStyled-sc-rqe5sa-0 ViewStyled__BorderWrapperStyled-sc-rqe5sa-4 kuBhDh eWayqq"
      data-testid="select"
      width="100%"
    >
      <div
        class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 dUEIVo brcjID"
        color="#203F59"
        data-testid="inputIcon"
        display="block"
      >
        <div
          class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 fTsvHV kHphsF"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="icon"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <input
          class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
          data-testid="iconInput"
          placeholder="Select Biller"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
